compiler/ilgen/IlBuilder.cpp

```diff
@@ -1419,6 +1419,77 @@ IlBuilder::Call(const char *functionName, int32_t numArgs, TR::IlValue ** argVal
    return NULL;
    }
 
+/** \brief
+ *     The service is used to atomically increase memory location [\p address + \p offset] by amount of \p value. 
+ *
+ *  \param address
+ *     The starting point of address we are trying to increment.
+ *
+ *  \param offset
+ *     The offset from address.
+ *
+ *  \param value
+ *     The amount of value we are trying to increment for the location.
+ * 
+ *  \note
+ *	   This service is created specifically for Int32.
+ *
+ *  \return
+ *     The old value at the location [\p address + \p offset] after increment.
+ */
+TR::IlValue *
+IlBuilder::AtomicInt32AddWithOffset(TR::IlValue * address, TR::IlValue * offset, TR::IlValue * value)
```

Comment:
Traditionally, you would not have a type-specific operation in JitBuilder. For example "Add" will add any pair of types. I can understand this code only supports Int32 right now, but the services should be AtomicAdd and AtomicAddWithOffset. You're taking the type from "value" anyway.